### ISSUE TITLE
Graph neural network-inspired kernels for gaussian processes in semi-supervised learning

### ISSUE THREAD

**Issue by 2nazero** [AIML-K/GNN_Survey#7]
[Graph neural network-inspired kernels for gaussian processes in semi-supervised learning](https://arxiv.org/abs/2302.05828)
```bib
@article{niu2023graph,
  title={Graph neural network-inspired kernels for gaussian processes in semi-supervised learning},
  author={Niu, Zehao and Anitescu, Mihai and Chen, Jie},
  journal={arXiv preprint arXiv:2302.05828},
  year={2023}
}
```

**Comment by 2nazero** [comment]
## Main Idea - The combination of GNNs and GPs
- There is a GNN model called Graph Convolutional Networks (GCN). The paper shows that when GCN layers have a large number of nodes, the GCN behaves similarly to a Gaussian Process.
- The authors created new kernels (functions that define relationships) for GPs that are inspired by GNNs.

**Comment by 2nazero** [comment]
## Scalable Computation through Low-Rank Approximation

- **Problem**: Computing the full covariance matrix $K^{(L)}$ recursively using Equations (3) and (4) is computationally expensive for large datasets.
- **Solution**: Use a **low-rank approximation** of $K^{(L)}$ to reduce computational costs. The idea is to approximate $K^{(L)}$ as $\hat{K}^{(L)} = Q^{(L)} Q^{(L)T}$, where $Q^{(L)}$ is a smaller matrix derived using **landmark nodes**.

### Key Approach
- **Approximation**: Start with a low-rank approximation of $C^{(l)}$ at each layer $l$ and use it to update $K^{(l+1)}$. 
- **Algorithm**: Algorithm 1 describes how $K^{(L)}$ is approximated using $Q^{(l)}$, which is computed iteratively to maintain efficiency.
<img width="667" alt="스크린샷 2024-10-23 오후 5 54 22" src="https://github.com/user-attachments/assets/6efdfbdd-15c3-4bf3-ae35-34c199e2f24d">


### Goal
- **Reduce Complexity**: This low-rank approach enables scalable computation by working with smaller matrices, making the GP-based model feasible for large-scale datasets.
<img width="627" alt="스크린샷 2024-10-23 오후 5 54 43" src="https://github.com/user-attachments/assets/7882fb20-5242-4fc2-8ee2-b3ec98a04ab2">

**Comment by 2nazero** [comment]
# Section 5: Composing Graph Neural Network-Inspired Kernels

- **Purpose**: This section explains how the **covariance matrix** of the Gaussian Process (GP) model derived from a Graph Convolutional Network (GCN) can be **programmatically composed** to match different GNN architectures.
  
- **Key Idea**: Using a **programmable procedure**, the covariance matrix and its low-rank approximation can be automatically transformed in the same way as the operations in a GNN. This allows creating corresponding **kernels** for various GNN architectures like **GCNII, GIN, and GraphSAGE**.

- **Transformation Process**:
  - For each GNN operation (like adding bias, multiplying weights, or applying an activation function), a corresponding transformation is applied to the covariance matrix.
  - For instance:
    - The graph convolution step modifies the covariance matrix from $K$ to $AK A^T$.
    - Multiplying weights transforms the covariance matrix as $K \to \sigma_w^2 K$.
  - These transformations are composed step-by-step to derive a **kernel** that aligns with the structure of a specific GNN layer.

- **Practical Examples**: 
  - The section shows specific compositions for **GCNII** and **GIN** layers, where the updates to the covariance matrix $K$ and low-rank factor $Q$ are defined to reflect the corresponding GNN operations.

- **Main Takeaway**: By treating the transformations on the covariance matrix in a programmable way, the paper demonstrates how **any GNN architecture** can be associated with a corresponding **GP kernel**, making it possible to extend the GP model to new GNN designs.

**Comment by 2nazero** [comment]
##Experiments

### Why was this experiment conducted?
The experiments aim to demonstrate the **performance of GP kernels** derived by taking limits on the layer width of GCNs and other GNN architectures. The primary goal is to show that these GPs are **comparable in prediction accuracy** while being **significantly faster** to compute.

### Datasets
The experiments were conducted on multiple benchmark datasets, covering both **classification** and **regression** tasks:
- **Citation Networks**: Cora, Citeseer, PubMed, ArXiv.
- **Community Detection**: Reddit.
- **Traffic Prediction**: Chameleon, Squirrel, and Crocodile (using Wikipedia hyperlinks).

### Evaluation Setup
1. **Prediction Performance (GCN-Based Comparison)**:
   - The GP kernels used include:
     - **GCNGP**: Derived from the limiting case of GCN.
     - **RBF**: A standard squared-exponential kernel.
     - **GGP**: A Gaussian Process kernel from related literature.
   - Each of these kernels has a **low-rank version** (indicated by `-X`), which uses a Nyström approximation for scalability.

2. **Performance Metrics**:
   - **Classification**: Micro-F1 score.
   - **Regression**: $R^2$ (coefficient of determination).

3. **Experiment 1: Comparing GCNs and GPs**:
   - The results show that **GCNGP** performs comparably or slightly better than GCN and other GP kernels. The **low-rank version** (GCNGP-X) achieves competitive results while being more computationally efficient.

4. **Experiment 2: Comparing with Other GNNs**:
   - The study extends the comparison to popular GNN architectures such as **GCNII, GIN, and GraphSAGE**.
   - The results show that the GP versions of these architectures perform similarly to the GNNs, but with efficiency advantages in specific cases like PubMed and Reddit.
<img width="671" alt="스크린샷 2024-10-23 오후 6 10 24" src="https://github.com/user-attachments/assets/45ae5a44-2b41-4250-8d2c-ba52eb11a833">


### Running Time and Scalability
1. **Running Time Comparison**:
   - The comparison shows that GCNGP-X is generally faster than GCN, with some datasets showing a speedup by one to two orders of magnitude. This is illustrated in **Figure 1**.
  
2. **Scalability Analysis**:
   - **Figure 2** demonstrates that the running time scales approximately linearly with respect to the graph size ($M + N$) for GCNGP-X and GCN. However, GCNGP shows cubic scaling, making the low-rank approximation more practical for large graphs.

3. **Analysis on the Depth**:
   - The study examines the impact of increasing the number of layers. The results (Figure 3) indicate that both GCN and GCNII suffer from **oversmoothing** at larger depths. However, their GP counterparts (e.g., GCNGP) remain stable even for a depth as large as 12.

4. **Analysis on the Landmark Set**:
   - The number of **landmark nodes** ($N_a$) affects the trade-off between approximation quality and running time. **Figure 4** shows that using only $1/800$ of the training set as landmarks achieves accuracy comparable to GCN, while the computational cost remains much lower.

### Results
- The experiments confirm that **GCNGP-X** performs competitively with GCNs in both accuracy and running time. The low-rank approximations enable scalability without sacrificing performance, making them viable for large-scale applications.
- The depth analysis also suggests that the GP-based models are less prone to **oversmoothing**, maintaining stable accuracy across multiple layers.
<img width="675" alt="스크린샷 2024-10-23 오후 6 09 39" src="https://github.com/user-attachments/assets/8b8ac3a0-b259-42ee-b7b4-91e5320ff7f5">


### Key Takeaway
The experiments show that GP kernels derived from GCNs can achieve **comparable prediction performance** to GNNs while being **computationally more efficient**. This demonstrates the practicality of the proposed GP approach for large-scale and deep graph-based tasks.

**Comment by 2nazero** [comment]
I have reviewed this paper but found it challenging to fully comprehend the mathematical proofs and rigorous details presented in the equations. If anyone could provide insights or comments, I would greatly appreciate it :)

**Comment by d-h-lee** [comment]
> I have reviewed this paper but found it challenging to fully comprehend the mathematical proofs and rigorous details presented in the equations. If anyone could provide insights or comments, I would greatly appreciate it :)

to fully comprehend the mathematical proofs --> as you clearly mentioned in https://github.com/AIML-K/GNN_Survey/issues/7#issuecomment-2430977827 , understanding NN and GP equivalence will help you understanding this. In particular, paper on infinitely-wide NN equivalent to GP will be useful. 

Don't go to the first source (Radford M Neal. Priors for infinite networks. In Bayesian Learning for Neural Networks, pages
29–53. Springer, 1996. -- too statistical). I recommend https://arxiv.org/abs/1711.00165 or https://openreview.net/pdf?id=rkl4aESeUH .

Meanwhile, your decision to attend GP workshop will pay off.

**Comment by 2nazero** [comment]
> I'd like to explore some extension questions related to this paper.
> 
> 1.In many real-world scenarios, graph-structured data changes dynamically over time (such as social networks). Can GNN-inspired GP kernels effectively handle such dynamic changes in graph structures?
> 
> 2.GNNs often encounter the 'over-smoothing' problem in deep architectures, where node representations tend to become similar as the number of layers increases, making it difficult for the model to distinguish features of different nodes. Can GNN-inspired GP methods avoid or reduce this over-smoothing issue? Is it possible to extend this approach to deeper graph structures while maintaining predictive performance and reducing information loss?

<img width="739" alt="스크린샷 2024-10-28 오후 6 02 13" src="https://github.com/user-attachments/assets/a4cd0f06-94be-4861-ac09-407107f7df79">

as you can see in the picture, covariance depends on x and x' which are the embeddings of the nodes, and if you look at the equation(1) I explained above, it changes when the node info changes.
